### ISSUE TITLE
Unused unexported methods/others.

### ISSUE THREAD

**Issue by albf** [dominikh/go-tools#785]
Hi there dominikh,

As a side project, I developed a checker that seems to go a bit further regarding unexported methods/values than what's currently present in staticcheck. What it identifies (all unexported definitions):

- global constants and variables (type assertions are ignored).
- structs and fields (assign-only is considered non-used).
- interfaces and its methods.
- functions, its arguments and return values (except main/init and cgo functions).
- methods and, as functions, arguments/return values (methods that match interfaces are ignored).
- functions/methods that are used as a value (therefore hard to detect) are also ignored to avoid false positives.

Current version uses SSA analysis and I believe at least the last two are not present - checking for argument usage or return values. Is there any interest in adding such a feature? I've been using it for a while in some large prod code bases and the results seem very useful.

Cheers,

Alex

**Comment by dominikh** [comment]
This basically sounds like #124 and #78. In other words, flagging unused parameters and return values – in unexported functions – has been on my list of things to look at. I'm undecided on whether checking for these things is actually a good idea, but I'd need to implement the check and run it against real code to figure that out.

It sounds like you put some thought into your implementation, avoiding false positives. If you'd like to contribute your code to Staticcheck, I'd be more than happy to look at it. I don't know, however, how straightforward contributing it would be. It seems unlikely that the architecture of your and my analysis match. What are your thoughts? Our implementation of the check lives at https://github.com/dominikh/go-tools/tree/master/unused

Out of curiosity, what prompted you to write your own analysis? The other points in your list seem to overlap heavily with what Staticcheck already checks for (though possibly avoiding outstanding issues such as #507 and #288)

**Comment by albf** [comment]
@dominikh, that's exactly like #124 and #78 (if we remove the overlap, of course). I believe my current code doesn't have #507 and #288 but mainly because it's using [buildssa](https://godoc.org/golang.org/x/tools/go/analysis/passes/buildssa), as opposed to buildir. I'm assuming you have a preference towards it, i.e. not using buildssa?

Architecture wise, my implementation is very straightforward: get buildssa/pass and identify usages/patterns. You can [take a look](https://github.com/albf/go-tools/tree/master/unexport) if you want. I've been using it against real code for a while and it's getting hard to find false positives - tests illustrate a bit the cases I ended up finding. Any thoughts regarding it being useful or not? Or about how to proceed?

As to why I wrote it, it was a little side project :smile:, noticed these patterns weren't being identified in my project and tried to do one. The company I work for supports staticcheck (it's a nice project) and suggested trying to contact you about it.

**Comment by dominikh** [comment]
That you don't suffer from #507 but I do is probably because you don't look at init functions, thus never see those field uses, while we currently do.

I've looked at your code, and it looks straightforward enough. It'll encounter some false negatives because it treats "used" as a single bit of information, not taking into account that usage information is a graph. For example, consider this package:

```
package pkg

func fn1() {
	fn1()
}
```

fn1 will be considered used, because fn1 uses fn1. You'll run into similar issues with mutual recursion.

Of course those are just false negatives, and your solution probably still finds 80% of unused code.

You do have a false positive due to tests, though. If a function is declared in a `foo.go` and only used in `foo_test.go`, then you'll miss that usage and mark the function unused. In Go's build system, what looks like one package really describes multiple packages. In this instance, the package made up of `foo.go`, and the test variant made up of `foo.go` and `foo_test.go`. Things get even worse once you consider external tests and build tags.

> Any thoughts regarding it being useful or not? Or about how to proceed?

To be honest, I'm not sure that much, if any, of the code can be adopted without major changes. You're operating on a different model than `unused`, and I think switching models would require a rewrite of your code. I do appreciate you reaching out and offering up your code, but I think it'd be more straightforward for me to fix #127 and #78 myself :-/

> As to why I wrote it, it was a little side project

As good a reason as any :D

**Comment by albf** [comment]
Good point about recursion. Regarding tests, I noticed that at some point but thought it would be a decent trade-off - if an unexported function is only used within tests, some might think it's fair to say it should be moved? Given the build system we use, this case is probably impossible to identify, so it's indeed a limitation. Also, it does check for init functions, but intentionally ignores assigns.

Agreed regarding major changes being required. Your unused code is not too large, if the strategy looks any good and you want any help trying to implement, feel free to ping me, I might be able to give it a shot.

Otherwise, thanks for the comments!
